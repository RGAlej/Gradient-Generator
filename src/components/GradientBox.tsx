import React from 'react';
import styled from 'styled-components';
import { HiTrash } from 'react-icons/hi';

import { media, setBorder, setColor, setTransition } from '../styles/styles';
import { useFunctionalContext } from '../contexts/functional.context';
import ColorList from './ColorList';

const GradientBox: React.FC = () => {
  const {
    gradient,
    codeToCopy,
    showTrash,
    dragOverHandler,
    dragLeaveHandler,
    dropHandler,
  } = useFunctionalContext();

  return (
    <Wrapper>
      <Bg
        style={{ background: gradient!.slice(0, -1) }} // to avoid the creation of a lot of classes for each gradient I use inline style
        data-testid='box'
      />
      <ColorList />
      <textarea ref={codeToCopy} value={gradient} readOnly />
      <HiTrash
        data-testid='trash'
        className={showTrash ? 'trash display' : 'trash'}
        onDragOver={(e) => dragOverHandler!(e)}
        onDragLeave={(e) => dragLeaveHandler!(e)}
        onDrop={(e) => dropHandler!(e)}
      />
    </Wrapper>
  );
};

const Wrapper = styled.main`
  display: grid;
  grid-template-columns: 1fr;
  grid-template-rows: 1fr repeat(2, min-content) 0.15fr;
  place-items: center;
  height: 90vh;
  background: ${(props) => props.theme.color1};

  textarea {
    outline: none;
    font-family: inherit;
    font-size: 1.5rem;
    font-weight: 600;
    text-align: center;
    color: ${(props) => props.theme.color3};
    width: 90%;
    padding-top: 4rem;
    padding-bottom: 10rem;
    border: none;
    background: transparent;
    resize: none;

    ::selection {
      background: transparent;
    }
  }

  .trash {
    fill: ${(props) => props.theme.color1};
    height: 0;
    width: 0;
    padding: 1rem;
    ${setBorder({ style: 'none' })};
    background: ${(props) => props.theme.color3};
    box-shadow: 0.1rem 0.1rem 1rem ${(props) => props.theme.color3};
    opacity: 0;
    visibility: hidden;
    ${setTransition()};
  }

  .display {
    height: 5rem;
    width: 5rem;
    opacity: 1;
    visibility: visible;
  }

  .dragging {
    fill: ${setColor.roseColor};
    height: 6rem;
    width: 6rem;
    background: ${(props) => props.theme.color2};
  }
`;

const Bg = styled.article`
  grid-row: 1 / span 1;
  height: 90%;
  width: 60%;
  ${setBorder({ style: 'none' })};
  box-shadow: 0.1rem 0.1rem 0.5rem ${(props) => props.theme.color3};

  ${media.phone`
    height: 60%;
    width: 85%;
  `}
`;

export default GradientBox;
