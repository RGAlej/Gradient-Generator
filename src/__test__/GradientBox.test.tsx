import {
  fireEvent,
  renderWithRouter,
  screen,
} from './utils/testing_library_utils';
import { RouteModel } from '../models/ManagementState.model';

import GradientBox from '../components/GradientBox';

describe('Box element', () => {
  test('check if all gradient box is correctly rendered', async () => {
    renderWithRouter(<GradientBox />, { route: RouteModel.GRADIENT });
    const gradientBox = screen.getByTestId('box');
    expect(gradientBox).toBeInTheDocument();

    // check color list
    const colorList = screen.getAllByTestId('color');
    expect(colorList).toHaveLength(2);

    const color1 = colorList[0] as HTMLInputElement;
    expect(color1).toHaveValue('#a1ffce');

    // check text area and default linear gradient
    const codeTextArea = screen.getByRole('textbox');
    expect(codeTextArea).toHaveTextContent('linear-gradient');
    expect(codeTextArea).toHaveTextContent('#A1FFCE');
  });

  test('check if change color works', async () => {
    renderWithRouter(<GradientBox />, { route: RouteModel.GRADIENT });
    const colorList = screen.getAllByTestId('color');
    const color1 = colorList[0] as HTMLInputElement;
    const codeTextArea = screen.getByRole('textbox');

    // check default values
    expect(color1).toHaveValue('#a1ffce');

    expect(color1).toMatchSnapshot();
    expect(codeTextArea).toMatchSnapshot();

    // change value of color1 input
    fireEvent.change(color1, { target: { value: '#ffffff' } });
    expect(color1).toHaveValue('#ffffff');

    // check if the code in text area was updated
    expect(codeTextArea).toHaveTextContent('linear-gradient');
    expect(codeTextArea).toHaveTextContent('#FFFFFF');

    expect(color1).toMatchSnapshot();
    expect(codeTextArea).toMatchSnapshot();
  });

  test('check trash status', async () => {
    renderWithRouter(<GradientBox />, { route: RouteModel.GRADIENT });
    const colorList = screen.getAllByTestId('color');
    const color1 = colorList[0] as HTMLInputElement;

    // trash should be hidden in the document
    const trash = await screen.findByTestId('trash');
    expect(trash).toHaveClass('trash');

    fireEvent.drag(color1);

    // trash should not be visible in the document if color list is less than three in length
    const trashDragged = await screen.findByTestId('trash');
    expect(trashDragged).not.toHaveClass('display');
  });
});
