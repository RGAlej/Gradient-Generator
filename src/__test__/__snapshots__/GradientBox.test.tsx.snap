// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Box element check if change color works 1`] = `
<input
  class="sc-gsDJrp cAoqju"
  data-testid="color"
  draggable="false"
  type="color"
  value="#A1FFCE"
/>
`;

exports[`Box element check if change color works 2`] = `
<textarea
  readonly=""
>
  linear-gradient(60deg, #A1FFCE, #FAFFD1);
</textarea>
`;

exports[`Box element check if change color works 3`] = `
<input
  class="sc-gsDJrp cAoqju"
  data-testid="color"
  draggable="false"
  type="color"
  value="#ffffff"
/>
`;

exports[`Box element check if change color works 4`] = `
<textarea
  readonly=""
>
  linear-gradient(60deg, #FFFFFF, #FAFFD1);
</textarea>
`;
